### ISSUE TITLE
[Improve Concept Docs]: `functools` Module in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3098]
This issue describes desired additions and improvements to the `functools` module **concept docs.**.
The related **concept exercise** issue and specs can be found [here](https://github.com/exercism/python/issues/3097),

<br>

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.  Additionally, we recommend reading the existing `about.md` docs [here](https://github.com/exercism/python/blob/main/concepts/functools/about.md), and an example of completed concept docs [here](https://github.com/exercism/python/tree/main/concepts/classes).

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🙏🏽   Desired Improvements and Changes

Generally, re-read the documents from a student perspective, improving tone, explanations, and examples.  Currently, the documents read as if they are documentation, and not teaching docs.

- [ ]   Clean up `introduction.md`, so that it only _introduces_ the module, and gives reasons why one would use the tools.  It's also OK to list the tools that will be covered, but not to go into detail.  Please also remove the function signatures and highlighting.  Remember:  this is the doc a student sees **before** going through the exercise, so it is best not to bog them down with too much detail.
- [ ]  Add more about what the module does, and why one would want to use the tools in the module to both the `about.md` and the `introduction.md`
- [ ]  Add real links to `links.json` for further student study/learning
- [ ]  Remove the highlighted function signatures from the doc, and re-work the text to be more narrative and focused on usage and explanation, rather than reading like a copy of the Python documentation.
- [ ]  Provide better use-cases and real-world code examples for the functions covered.
- [ ]  Provide better use-case, explanation and examples for `update_wrapper` and `wraps`.  If this feels forced, remove the methods and let a maintainer know.  We may move these to the `decorators` concept.

___________

# 🌟   Original Specifications

Below are the original specifications from the Concept issue.  Please use these as a guide in addition to the change list above.

This issue describes how to implement the `functools` module **concept docs.**.
You can find the related **concept exercise** issue [here](https://github.com/exercism/python/issues/3097)

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯   Goal

These concept docs are meant to teach an understanding/use of the `functools` module in Python.

<br>

## 💡   Learning objectives

Learn more about the functional tools the Python Standard Library provides through the `functools` module.
Build and understanding of and use the following methods and decorators from the module:

- `functools.partial()`
  - `partial.func`
  - `partial.args`
  - `partial.keywords`
-   *class* `functools.partialmethod()` 
-  *class* `functools.singledispatchmethod()`
-  `functools.cache()`
- `@functools.lru_cache()`
- `@functools.singledispatch`
  - `@<function>.register()`
- `@functools.wraps()` -- (_these might move to the `decorators` exercise, depending on how difficult they are to work in_)
   - `functools.update_wrapper()`
-  ~~`@functools.cached_property()`~~ (_this seems better in `class_customization`_)
-  ~~`@functools.total_ordering`~~ (_this method seems more appropriate in the`rich comparisons` exercise_)

<br>

## 🚫  Out of scope

<details>
<summary><b>Concepts and Subjects that are Out of Scope</b></summary>
<br>

- `classes`  & `class customization` beyond the direct use of the _class_ methods in this module.
- `comprehensions`
- `comprehensions` in `lambdas`
-  `decorators` (_these have their own exercise.  See issue #2356_ )
- `map()`, `filter()` or `functools.reduce()` in a `comprehension`
-  `functools.reduce()`(_this was already covered with `map()` and `filter()`_)
- `generators`, beyond what is in this module
-  using an `assignment expression` or "walrus" operator (`:=`) in a `lambda`
- class decorators beyond the ones described in this module.
- `enums`

</details>

<br>

## 🤔  Concepts

<details>
<summary><b>Concepts/Related Concepts Under this Topic</b></summary>
<br>

- functional tools in python
- `functools` module
- `generic functions`
- `decorators`
- `higher-order functions`
- `partial objects` in python/`partial evaluation` in python
- `single dispatch`

</details>

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs</summary>
<br>

- `basics`
- `bools`
- `classes`
- `class-customization`
- `class-components`
- `comparisons`
- `rich-comparisons`
- `decorators`
- `descriptors`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order-functions`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>  

## 📚  Resources to refer to

<details>
<summary><b>Resources</b></summary>
<br>

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs: functional programming HOWTO](https://docs.python.org/3/howto/functional.html)
- [Python Docs: functools module](https://docs.python.org/3/library/functools.html)
- [Pthon Module of the Week:  `functools` - Tools for Maniputlating Fuctions](https://pymotw.com/3/functools/)
- [Florian Dahlitz: Introduction to Python's Functools Module](https://florian-dahlitz.de/blog/introduction-to-functools)
- [PyDanny: Python Partials are Fun!](https://www.pydanny.com/python-partials-are-fun.html)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Trey Hunner:  Ist it a Class or a Function?  It's a callable!](https://treyhunner.com/2019/04/is-it-a-class-or-a-function-its-a-callable/)
- [built-ins: Python Docs](https://docs.python.org/3/library/functions.html)
  - [`map()`](https://docs.python.org/3/library/functions.html#map) & [`filter()`](https://docs.python.org/3/library/functions.html#filter)
  - [`any()`](https://docs.python.org/3/library/functions.html#any) & [`all()`](https://docs.python.org/3/library/functions.html#all)
  - [`min()`](https://docs.python.org/3/library/functions.html#min) & [`max()`](https://docs.python.org/3/library/functions.html#max)
  - [`sum()`](https://docs.python.org/3/library/functions.html#sum) 
- [Real Python: Functional Programming in Python: When and How to Use it](https://realpython.com/python-functional-programming/)

</details>
<br>

##  📁  Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
@Metallifax  -- This is the "improve concept docs.  Comment to be assigned.  😄

**Comment by Metallifax** [comment]
:wave: :taco:

**Comment by Metallifax** [comment]
:wave: Comment!

**Comment by BethanyG** [comment]
@Metallifax -

Evidently, the bot is _anxious_   🙄   Will fix that up tomorrow my time. Apologies for the ping.

**Comment by BethanyG** [comment]
@Metallifax  -- Checking in here .. do you still want to work on this?  Anything I can do to help?
